src/js/tooltip.js

```diff
@@ -0,0 +1,420 @@
+		}
+
+		// Check that the value of tooltip position is valid. Default to below.
+		if (opts.tooltipPosition) {
+			if (Tooltip.validTooltipPositions.indexOf(opts.tooltipPosition) === -1) {
+				Tooltip.throwError("Invalid value for tooltip position. Valid values are:" + Tooltip.validTooltipPositions.toString() +" or nothing which will default to a value of `below`");
+			}
+		} else {
+			opts.tooltipPosition = "below";
+		}
+
+		return opts;
+	};
+
+	/**
+	 * Render the tooltip. Adds markup and attributes to this.tooltipEl in the DOM
+	*/
+	render() {
+
+		this.tooltipEl.setAttribute('role', 'tooltip');
+		this.tooltipEl.classList.add('o-tooltip');
+
+		if (this.opts.zIndex) {
+			this.tooltipEl.style.zIndex = this.opts.zIndex;
+		}
+
+		// Build and append the close button
+		const button = document.createElement('a');
+		button.className = 'o-tooltip-close';
+		button.setAttribute('role', 'button');
+		button.setAttribute('href', '#void');
+		button.setAttribute('aria-label', 'Close');
+		button.setAttribute('title', 'Close');
+		this.tooltipEl.appendChild(button);
+
+	};
+
+	/**
+	 * Show the tooltip. Adds event handlers for clicks, touches, keypresses and
+	 * viewport resizes. Uses FTDomDelegate to implement the event delegate
+	 * pattern. Calls DrawTooltip.
+	*/
+	show() {
+		// Delegate pattern
+		this.delegates.doc.root(document.body);
+		this.delegates.tooltip.root(this.tooltipEl);
+
+		// Set up all the ways to close the tooltip
+		this.closeHandler = this.close.bind(this);
+		this.delegates.tooltip.on('click', '.o-tooltip-close', this.closeHandler);
+		this.delegates.tooltip.on('touchend', '.o-tooltip-close', this.closeHandler);
+
+		this.closeOnExternalClickHandler = this.closeOnExternalClick.bind(this);
+		this.delegates.doc.on('click', 'body', this.closeOnExternalClickHandler);
+		this.delegates.doc.on('touchend', 'body', this.closeOnExternalClickHandler);
+
+		this.closeOnKeyUpHandler = this.closeOnKeyUp.bind(this);
+		this.delegates.doc.on('keyup', this.closeOnKeyUpHandler);
+
+		// Resize means we might need to move our tooltip so it is still on screen
+		this.resizeListenerHandler = this.resizeListener.bind(this);
+		this.delegates.doc.on('oViewport.resize', 'body', this.resizeListenerHandler);
+
+		this.drawTooltip();
+		this.visible = true;
+	};
+
+	/**
+	 * Destroy the tooltip.
+	*/
+	destroy() {
+		if (this.visible === true) {
+			this.close();
+		}
+
+		Tooltip._tooltips.delete(this.tooltipEl);
+		if (Tooltip._tooltips.size === 1) {
+			delete Tooltip._tooltips;
+		}
+	};
+
+	/**
+	 * Close the tooltip. (Visually hide it and remove event listeners)
+	*/
+	close() {
+		this.delegates.doc.destroy();
+		this.delegates.tooltip.destroy();
+
+		//Viewport.stopListeningTo('resize');TODO <- this should only stop listening if it's the only tooltip on the page
+		this.visible = false;
+		this.tooltipEl.style.display = 'none';
+		return false;
+	};
+
+	closeOnExternalClick() {
```

Comment:
As this method is a no-op, I don't think we should be adding three event listeners which trigger it